lib/mongo/server/connection_pool/queue.rb

```diff
@@ -165,12 +164,45 @@ def wait_timeout
           @wait_timeout ||= options[:wait_queue_timeout] || WAIT_TIMEOUT
         end
 
+        # The maximum seconds a socket can remain idle in the connection pool after it
+        #   has been checked out and checked back in.
+        #
+        # @example Get the max idle time.
+        #   queue.max_idle_time
+        #
+        # @return [ Float ] The max socket idle time in seconds.
+        #
+        # @since 2.5.0
+        def max_idle_time
+          @max_idle_time ||= options[:max_idle_time]
+        end
+
+        # Close sockets that have been open for longer than the max idle time, if the
+        #   option is set.
+        #
+        # @example Close the stale sockets
+        #   queue.close_stale_sockets!
+        #
+        # @since 2.5.0
+        def close_stale_sockets!
+          if max_idle_time
+            mutex.synchronize do
+              queue.reverse_each do |connection|
+                if last_checkin = connection.last_checkin
+                  connection.disconnect! if (Time.now - last_checkin) > max_idle_time
+                end
+                break if queue.index(connection) == min_size
```

Comment:
Yes, I saw that PyMongo does that but I didn't do it because it would require doing two loops through all connections while the lock is held. I don't know if the benefit of closing, then reconnecting the min number of sockets is worth it? Maybe it is up to the user then to make sure min_size isn't ridiculously high so that this method doesn't hold the lock for so long?